services/m3aggregator/processor/processor.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package processor
+
+import (
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/m3db/m3metrics/metric/aggregated"
+	"github.com/m3db/m3metrics/policy"
+	"github.com/m3db/m3metrics/protocol/msgpack"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/sync"
+)
+
+var (
+	errProcessorQueueIsFull = errors.New("processor queue is full")
+	errProcessorIsClosed    = errors.New("processor is closed")
+)
+
+// MetricWithPolicyFn handles aggregated metrics with corresponding policy.
+// It must be thread-safe.
+type MetricWithPolicyFn func(metric aggregated.Metric, policy policy.Policy) error
+
+// AggregatedMetricProcessor processes aggregated metrics
+type AggregatedMetricProcessor struct {
+	sync.RWMutex
+
+	closed    bool
+	queue     chan msgpack.BufferedEncoder
+	workers   xsync.WorkerPool
+	wgWorkers sync.WaitGroup
+	fn        MetricWithPolicyFn
+	log       xlog.Logger
+	iterOpts  msgpack.AggregatedIteratorOptions
+}
+
+// NewAggregatedMetricProcessor creates a new aggregated metric processor
+func NewAggregatedMetricProcessor(opts Options) *AggregatedMetricProcessor {
```

Comment:
I'm not sure I like the idea of unbounded memory growth under high load. It feels better that we preallocate enough space for the channel based on the total memory capacity and the number of processes running on the server, and just have the process always running in the steady state without having its memory usage shrinking or expanding a ton. If a particular process is under load, that process should take measures to alleviate pressure on itself (e.g., by dropping old metrics in queue). If we allow unlimited memory growth without shedding load, a badly behaving process might use a ton of memory, negatively affecting the performance of its good neighbors on the same box. FWIW this is the same approach we adopt in shunts.